src/main/java/uk/ac/ebi/eva/contigalias/datasource/ENAAssemblyDataSource.java

```diff
@@ -0,0 +1,108 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.ac.ebi.eva.contigalias.datasource;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+
+import uk.ac.ebi.eva.contigalias.dus.ENAAssemblyReportReader;
+import uk.ac.ebi.eva.contigalias.dus.ENAAssemblyReportReaderFactory;
+import uk.ac.ebi.eva.contigalias.dus.ENABrowser;
+import uk.ac.ebi.eva.contigalias.dus.ENABrowserFactory;
+import uk.ac.ebi.eva.contigalias.entities.AssemblyEntity;
+import uk.ac.ebi.eva.contigalias.entities.ChromosomeEntity;
+import uk.ac.ebi.eva.contigalias.entities.ScaffoldEntity;
+import uk.ac.ebi.eva.contigalias.entities.SequenceEntity;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Repository("ENADataSource")
+public class ENAAssemblyDataSource implements AssemblyDataSource {
+
+    private final ENABrowserFactory factory;
+
+    private final ENAAssemblyReportReaderFactory readerFactory;
+
+    @Autowired
+    public ENAAssemblyDataSource(ENABrowserFactory factory,
+                                 ENAAssemblyReportReaderFactory readerFactory) {
+        this.factory = factory;
+        this.readerFactory = readerFactory;
+    }
+
+    @Override
+    public Optional<AssemblyEntity> getAssemblyByAccession(String accession) throws IOException {
+        ENABrowser enaBrowser = factory.build();
+        enaBrowser.connect();
+
+        AssemblyEntity assemblyEntity;
+        try (InputStream stream = enaBrowser.getAssemblyReportInputStream(accession)) {
+            ENAAssemblyReportReader reader = readerFactory.build(stream);
+            assemblyEntity = reader.getAssemblyEntity();
+        } finally {
+            enaBrowser.disconnect();
+        }
+        return Optional.of(assemblyEntity);
+    }
+
+    /**
+     * Adds ENA sequence names to chromosomes and scaffolds in an assembly.
+     *
+     * @param optional {@link AssemblyEntity} to get ENA sequence names for
+     * @throws IOException Passes IOException thrown by {@link #getAssemblyByAccession(String)}
+     */
+    public void getENASequenceNamesForAssembly(Optional<AssemblyEntity> optional) throws IOException {
+        if (optional.isPresent()) {
+            AssemblyEntity targetAssembly = optional.get();
+            if (!hasAllEnaSequenceNames(targetAssembly)) {
+                String genbank = targetAssembly.getGenbank();
+                Optional<AssemblyEntity> enaAssembly = getAssemblyByAccession(genbank);
+
+                if (enaAssembly.isPresent()) {
+                    AssemblyEntity sourceAssembly = enaAssembly.get();
+                    putENASequenceNames(sourceAssembly.getChromosomes(), targetAssembly.getChromosomes());
```

Comment:
The method name is getSomething  which, in Java convention, must return something. We also invoke put inside that get method which is a bit confusing. **Judgment call**: Mutating arguments that are passed in to a method might not be a good idea. If we do intend to get a modified AssemblyEntity from this method, I suggest callling that out explicitly in the documentation.

Reply:
I've renamed both methods to addSomething which is hopefully clearer, and updated the documentation.